### ISSUE TITLE
Wrong drawing when applying (Identity) transform to Direct2D canvas 

### ISSUE THREAD

**Issue by jeromebjerome** [Microsoft/Win2D#170]
Hello,
Below is a very basic C++ Universal  App showing a difference between using Win2D and using direct access to Direct2D API.
Sample: I draw a line (thickness 1) between (0,0) and (100,100)
1)	With Win2D, we can see the line start precisely from the top left corner of the canvas
2)	 With Direct2D API, we can see that  the line starts from below the corner, shifted by several pixels
I could identify that the call:  pTarget->SetTransform(D2D1::Matrix3x2F::Identity()); just before DrawLine is causing the shift.
This problem leads to important issues on hit testing on drawn geometry.

Source:
//
// MainPage.xaml.cpp
// Implementation of the MainPage class.
//

#include <wrl.h>
#include <ppl.h>

#include "pch.h"
#include "MainPage.xaml.h"

#include <Microsoft.Graphics.Canvas.native.h>
#include <d2d1_2.h>
#include <collection.h>
#include <algorithm>

#include <stdio.h>
#include <time.h>   
#include <math.h>   

using namespace picking;

using namespace Windows::UI::Xaml;
using namespace Platform::Collections;
using namespace Windows::Foundation::Collections;
using namespace Microsoft::Graphics::Canvas;
using namespace std;
using namespace Microsoft::WRL;
using namespace Windows::ApplicationModel::Resources;
using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Core;
using namespace Windows::UI::Xaml::Controls;
using namespace Windows::UI::Xaml::Controls::Primitives;
using namespace Windows::UI::Xaml::Data;
using namespace Windows::UI::Xaml::Input;
using namespace Windows::UI::Xaml::Media;
using namespace Windows::UI::Xaml::Navigation;

// The Blank Page item template is documented at http://go.microsoft.com/fwlink/?LinkId=402352&clcid=0x409

MainPage::MainPage()
{
	InitializeComponent();
}

#define mytest

void picking::MainPage::grafic_zone_Draw(Microsoft::Graphics::Canvas::UI::Xaml::CanvasControl^ sender, Microsoft::Graphics::Canvas::UI::Xaml::CanvasDrawEventArgs^ args)
{
#ifdef mytest
	ComPtr<ID2D1RenderTarget> target = GetWrappedResource<ID2D1RenderTarget>(args->DrawingSession);
	ID2D1RenderTarget *pTarget = target.Get();
	if (pTarget) {

		ID2D1SolidColorBrush* m_MyBrosse = nullptr;
		pTarget->CreateSolidColorBrush(
			D2D1::ColorF(D2D1::ColorF::Red),
			&m_MyBrosse
			);

		pTarget->SetTransform(D2D1::Matrix3x2F::Identity());

		pTarget->DrawLine(
			D2D1::Point2F(0, 0),
			D2D1::Point2F(100, 100),
			m_MyBrosse);
	}
#else
	Windows::UI::Color couleur;
	couleur.A = 255;
	couleur.R = 255;
	couleur.G = 0;
	couleur.B = 0;

	args->DrawingSession->DrawLine(0, 0, 100, 100, couleur, 1);
#endif

}

void picking::MainPage::grafic_zone_PointerMoved(Platform::Object^ sender, Windows::UI::Xaml::Input::PointerRoutedEventArgs^ e)
{
	Windows::UI::Xaml::Input::Pointer ^ptr = e->Pointer;

	if (
		ptr->PointerDeviceType == Windows::Devices::Input::PointerDeviceType::Mouse ||
		ptr->PointerDeviceType == Windows::Devices::Input::PointerDeviceType::Touch) {

		// To get mouse state, we need extended pointer details.
		// We get the pointer info through the getCurrentPoint method
		// of the event argument. 
		//Windows::UI::Input::PointerPoint ^ptrPt = e->GetCurrentPoint(nullptr);
		Windows::UI::Input::PointerPoint ^ptrPt = e->GetCurrentPoint(grafic_zone);

		double X = ptrPt->Position.X;
		double Y = ptrPt->Position.Y;

		if (X < 100) {
			if ((X - Y) <= 2 && (X - Y) >= -2) {
				return;
			}
		}
	}

	// Prevent most handlers along the event route from handling the same event again.
	e->Handled = true;
}

![with_d2d](https://cloud.githubusercontent.com/assets/13735023/10244012/b6d03c4a-68fc-11e5-8483-1e266ea6ba22.jpg)
![with_win2d](https://cloud.githubusercontent.com/assets/13735023/10244013/b6d4b90a-68fc-11e5-869e-937defdb506a.jpg)

**Comment by shawnhar** [comment]
XAML uses texture atlasing behind the scenes, so when you use CanvasControl it is actually drawing to just part of a potentially larger output surface.  In native XAML+D2D model, the app is responsible for including an appropriate offset in all its drawing commands.

Win2D handles this for you by automatically including the necessary offset in its transform matrix, but if you reset that transform yourself from native D2D, it's up to you to preserve the offset while doing so.

You can determine the offset by looking up the previous transform that was set by Win2D.  If you do this from Win2D, you'll get identity (assuming you haven't set something different), but if you do it from native D2D, you'll see a translation matrix with the appropriate offset.

In short - to set the transform from native D2D, you should concatenate onto the existing translation, rather than overwriting it.

**Comment by jeromebjerome** [comment]
Hello,

Not sure to understand the very technical reason but I did as you said and
it worked for me.
Thank you.

On Fri, Oct 2, 2015 at 2:38 PM, Shawn Hargreaves <notifications@github.com>
wrote:

> XAML uses texture atlasing behind the scenes, so when you use
> CanvasControl it is actually drawing to just part of a potentially larger
> output surface. In native XAML+D2D model, the app is responsible for
> including an appropriate offset in all its drawing commands.
>
> Win2D handles this for you by automatically including the necessary offset
> in its transform matrix, but if you reset that transform yourself from
> native D2D, it's up to you to preserve the offset while doing so.
>
> You can determine the offset by looking up the previous transform that was
> set by Win2D. If you do this from Win2D, you'll get identity (assuming you
> haven't set something different), but if you do it from native D2D, you'll
> see a translation matrix with the appropriate offset.
>
> In short - to set the transform from native D2D, you should concatenate
> onto the existing translation, rather than overwriting it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Microsoft/Win2D/issues/170#issuecomment-145006265>.
>



-- 
jérôme.
